File: libs/py-common/windrose_common/projection.py
"""Granular authz-projection loader (MASTER-FR-012 / RBC-FR-040).

This mirrors ``libs/go-common/opaclient/projection.go``: it assembles the
``input.projection`` slice OPA's ``windrose.authz_input`` policy evaluates by
reading the *granular* ``perm:*`` projection keys that rbac-service's projector
actually writes (catalog, flags, tenant actions, workspace assignment, resource
grant, tenant meta). Go and Python services thus consume the identical rbac
projection — no separate pre-assembled per-(user,action,workspace) key is
needed.

This module is additive: it does not change the existing single-key
``OpaClient`` path. Callers that want the granular path build the projection
here and pass it to ``OpaClient.decision(projection=...)``.
"""

from __future__ import annotations

import hashlib
import json
from typing import Any

import redis.asyncio as aioredis

CATALOG_KEY = "perm:catalog:actions"


def urn_hash(urn: str) -> str:
    """Resource-grant key suffix: first 32 hex chars of sha256(urn). Matches
    rbac ``domain.URNHash`` and go-common ``opaclient.URNHash``."""
    return hashlib.sha256(urn.encode()).hexdigest()[:32]


def effective_user(subject: dict[str, Any]) -> str:
    """Whose projection is read: OBO calls resolve to the original user."""
    if subject.get("typ") == "agent_obo" and subject.get("obo_sub"):
        return str(subject["obo_sub"])
    return str(subject.get("id", ""))


async def _get_json(redis: aioredis.Redis, key: str) -> dict[str, Any] | None:
    raw = await redis.get(key)
    if not raw:
        return None
    try:
        return json.loads(raw)
    except (ValueError, TypeError):
        return None


async def load_projection(
    redis: aioredis.Redis,
    *,
    tenant: str,
    subject: dict[str, Any],
    action: str,
    workspace_id: str | None = None,
    resource_urn: str | None = None,
) -> dict[str, Any]:
    """Build the OPA ``input.projection`` from the granular ``perm:*`` keys."""
    user = effective_user(subject)
    ws = workspace_id or ""
    proj: dict[str, Any] = {
        "action_known": False,
        "action_scoped": False,
        "autonomous_enabled": False,
        "flags": {"found": False, "admin": False, "ws_admin": []},
        "tenant_actions": {"found": False, "actions": []},
        "workspace": {"assigned": False, "actions": [], "archived": False},
        "resource": {"found": False, "level": "", "archived": False},
        "workspace_archived_tenant": False,
    }

    # Catalog: action -> workspace_scoped (global key).
    cat = await _get_json(redis, CATALOG_KEY)
    if cat and isinstance(cat.get("actions"), dict):
        actions = cat["actions"]
        if action in actions:
            proj["action_known"] = True
            proj["action_scoped"] = bool(actions[action])

    # Flags.
    flags = await _get_json(redis, f"perm:{tenant}:{user}:flags")
    if flags is not None:
        proj["flags"] = {
            "found": True,
            "admin": bool(flags.get("admin", False)),
            "ws_admin": list(flags.get("ws_admin") or []),
        }

    # Tenant-scoped role actions.
    tact = await _get_json(redis, f"perm:{tenant}:{user}:actions")
    if tact is not None:
        proj["tenant_actions"] = {"found": True, "actions": list(tact.get("actions") or [])}

    # Workspace assignment for the request's workspace.
    if ws:
        wentry = await _get_json(redis, f"perm:{tenant}:{user}:ws:{ws}")
        if wentry is not None and not wentry.get("deleted"):
            proj["workspace"] = {
                "assigned": True,
                "actions": list(wentry.get("actions") or []),
                "archived": bool(wentry.get("archived", False)),
            }
        archived = await _get_json(redis, f"perm:{tenant}:archived_ws")
        if archived and ws in (archived.get("ws") or []):
            proj["workspace_archived_tenant"] = True

    # Resource grant overlay for the request's URN.
    if resource_urn:
        rentry = await _get_json(redis, f"perm:{tenant}:{user}:res:{urn_hash(resource_urn)}")
        if rentry is not None and not rentry.get("deleted"):
            proj["resource"] = {
                "found": True,
                "level": str(rentry.get("level", "")),
                "archived": bool(rentry.get("archived", False)),
            }

    # Tenant meta: autonomous-agent enablement.
    meta = await _get_json(redis, f"perm:{tenant}:meta")
    if meta is not None:
        proj["autonomous_enabled"] = bool(meta.get("autonomous_enabled", False))

    return proj

File: libs/go-common/opaclient/projection.go
package opaclient

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"

	"github.com/windrose-ai/go-common/redisx"
)

// ProjectionLoader reads the permissions_flat projection slice for one decision
// from Redis, using the rbac-service key scheme (RBC-FR-040). This is what a
// service's request path uses to feed OPA (MASTER-FR-012): O(1) key reads, no
// synchronous rbac call.
type ProjectionLoader struct {
	R *redisx.Client
}

// NewLoader builds a loader over a redisx client.
func NewLoader(r *redisx.Client) *ProjectionLoader { return &ProjectionLoader{R: r} }

// URNHash is the resource-grant key suffix: sha256(urn) hex, first 32 chars.
// Matches rbac domain.URNHash and the Rego crypto.sha256 substring.
func URNHash(urn string) string {
	sum := sha256.Sum256([]byte(urn))
	return hex.EncodeToString(sum[:])[:32]
}

// EffectiveUser resolves whose projection is read (OBO → original user).
func EffectiveUser(s Subject) string {
	if s.Typ == "agent_obo" && s.OboSub != "" {
		return s.OboSub
	}
	return s.ID
}

// Load builds the Projection facts for (tenant, action, workspace, resource)
// from Redis. Tombstoned subsidiary keys (deleted=true) read as absent.
func (l *ProjectionLoader) Load(ctx context.Context, in *Input) (Projection, error) {
	tenant := in.Tenant
	user := EffectiveUser(in.Subject)
	var p Projection

	// Catalog: action -> workspace_scoped (global key, no TTL).
	if raw, ok, err := l.R.Get(ctx, "perm:catalog:actions"); err != nil {
		return p, err
	} else if ok {
		var cv struct {
			Actions map[string]bool `json:"actions"`
		}
		if json.Unmarshal([]byte(raw), &cv) == nil {
			scoped, known := cv.Actions[in.Action]
			p.ActionKnown = known
			p.ActionScoped = scoped
		}
	}

	// Flags.
	if raw, ok, err := l.R.Get(ctx, key("perm:%s:%s:flags", tenant, user)); err != nil {
		return p, err
	} else if ok {
		var fv struct {
			Admin   bool     `json:"admin"`
			WsAdmin []string `json:"ws_admin"`
		}
		if json.Unmarshal([]byte(raw), &fv) == nil {
			p.Flags = Flags{Found: true, Admin: fv.Admin, WsAdmin: emptyIfNil(fv.WsAdmin)}
		}
	}

	// Tenant-scoped actions.
	if raw, ok, err := l.R.Get(ctx, key("perm:%s:%s:actions", tenant, user)); err != nil {
		return p, err
	} else if ok {
		var av struct {
			Actions []string `json:"actions"`
		}
		if json.Unmarshal([]byte(raw), &av) == nil {
			p.TenantActions = TenantActions{Found: true, Actions: emptyIfNil(av.Actions)}
		}
	}

	// Workspace entry for the request's workspace.
	if in.WorkspaceID != "" {
		if raw, ok, err := l.R.Get(ctx, fmt.Sprintf("perm:%s:%s:ws:%s", tenant, user, in.WorkspaceID)); err != nil {
			return p, err
		} else if ok {
			var wv struct {
				Actions  []string `json:"actions"`
				Archived bool     `json:"archived"`
				Deleted  bool     `json:"deleted"`
			}
			if json.Unmarshal([]byte(raw), &wv) == nil && !wv.Deleted {
				p.Workspace = WorkspaceFacts{Assigned: true, Actions: emptyIfNil(wv.Actions), Archived: wv.Archived}
			}
		}
		// Is this workspace archived at the tenant level (admin write block, BR-7)?
		if raw, ok, err := l.R.Get(ctx, fmt.Sprintf("perm:%s:archived_ws", tenant)); err != nil {
			return p, err
		} else if ok {
			var aw struct {
				Workspaces []string `json:"ws"`
			}
			if json.Unmarshal([]byte(raw), &aw) == nil {
				for _, w := range aw.Workspaces {
					if w == in.WorkspaceID {
						p.WorkspaceArchivedTenant = true
						break
					}
				}
			}
		}
	}

	// Resource grant for the request's URN.
	if in.ResourceURN != "" {
		h := URNHash(in.ResourceURN)
		if raw, ok, err := l.R.Get(ctx, fmt.Sprintf("perm:%s:%s:res:%s", tenant, user, h)); err != nil {
			return p, err
		} else if ok {
			var rv struct {
				Level    string `json:"level"`
				Archived bool   `json:"archived"`
				Deleted  bool   `json:"deleted"`
			}
			if json.Unmarshal([]byte(raw), &rv) == nil && !rv.Deleted {
				p.Resource = ResourceFacts{Found: true, Level: rv.Level, Archived: rv.Archived}
			}
		}
	}

	// Tenant meta: autonomous-agent enablement.
	if raw, ok, err := l.R.Get(ctx, fmt.Sprintf("perm:%s:meta", tenant)); err != nil {
		return p, err
	} else if ok {
		var mv struct {
			AutonomousEnabled bool `json:"autonomous_enabled"`
		}
		if json.Unmarshal([]byte(raw), &mv) == nil {
			p.AutonomousEnabled = mv.AutonomousEnabled
		}
	}

	return p, nil
}

// CheckWithRedis loads the projection from Redis then evaluates it against OPA
// — the full request-path decision (MASTER-FR-012). When the projection came
// back empty (a Redis miss — cold cache after a restart/failover/flush, not a
// genuine "no grant") and a fallback is configured (EnableMissFallback,
// RBC-FR-045), this re-checks against rbac-service's SQL ground truth instead
// of returning the empty-projection deny — see withMissFallback.
func (c *Client) CheckWithRedis(ctx context.Context, l *ProjectionLoader, in Input) (Decision, error) {
	p, err := l.Load(ctx, &in)
	if err != nil {
		return Decision{}, err
	}
	in.Projection = p
	dec, err := c.Check(ctx, in)
	if err != nil {
		return dec, err
	}
	return c.withMissFallback(ctx, in, dec), nil
}

func key(f, tenant, user string) string { return fmt.Sprintf(f, tenant, user) }

func emptyIfNil(s []string) []string {
	if s == nil {
		return []string{}
	}
	return s
}
